### PR TITLE
[serve] Update docs for v2 Deployments API

### LINKED ISSUE #14808: Update docs for HTTP ingress

### LINKED ISSUE #14809: Update docs for .deploy()

### PR DESCRIPTION
**Pull request by edoakes**
## Related issue number

Closes https://github.com/ray-project/ray/issues/14809
Closes https://github.com/ray-project/ray/issues/14808

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@architkulkarni @simon-mo PTAL, this is still somewhat rough but I've gotten up through the "Core APIs' page (going sequentially through the existing pages). There are a few places where I have TODOs to put in links that don't exist yet.

**Review by architkulkarni — COMMENTED** [review]
I wrote some nits.  Looks great!

File: python/ray/serve/examples/doc/imported_backend.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
+    def __call__(self, arg):
+        from my_module import my_model
+        self.model = my_model(arg)
 
-import_path = "my_module.MyFuncOrClass"
-serve.create_backend("imported", import_path, "input_arg")
-serve.create_endpoint("imported", backend="imported", route="/imported")
 
-print(requests.get("http://127.0.0.1:8000/imported").text)
+MyDeployment.deploy("/model_path.pkl")
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/examples/doc/imported_backend.py:13]: updated the args to be more clear

File: doc/source/serve/core-apis.rst [hunk 2]
```diff
@@ -4,109 +4,124 @@ Core APIs
...
+  # You can also use Deployment.options() to change options without redefining
+  # the class.
+  SimpleDeployment.options(num_replicas=2).deploy()
 
-We can also query the endpoint using the :mod:`ServeHandle <ray.serve.handle.RayServeHandle>` interface.
+
+If you want to avoid always re-deploying, this behavior can be controlled by providing a ``version`` string for the deployment in the decorator or ``Deployment.options()``.
+If a ``version`` string is provided, a rolling update of the replicas will *only* happen if the version is updated.
+Otherwise, the call to ``.deploy()`` will be a no-op.
+This can be useful when you have many deployments in a script and you only want to re-deploy one of them each time the script runs.
```
[the remark this answers is not in the document]
Comment (edoakes) [reply doc/source/serve/core-apis.rst:79]: Thanks :)

**Comment by edoakes** [comment]
Ok @architkulkarni @simon-mo I updated the rest of the docs, this should be ready for a full review.

@architkulkarni I updated `core-apis.rst` with your suggested changes if you want to skim through it quickly again.

File: doc/source/serve/core-apis.rst [hunk 3]
```diff
@@ -1,128 +1,150 @@
...
-Exposing a Backend
-==================
-
-While backends define the implementation of your request handling logic, endpoints allow you to expose them via HTTP.
-Endpoints are "logical" and can have one or multiple backends that serve requests to them.
-To create an endpoint, we simply need to specify a name for the endpoint, the name of a backend to handle requests to the endpoint, and the route and methods where it will be accesible.
-By default endpoints are serviced only by the backend provided to :mod:`create_endpoint <ray.serve.api.create_endpoint>`, but in some cases you may want to specify multiple backends for an endpoint, e.g., for A/B testing or incremental rollout.
-For information on how to do this, please see :ref:`serve-split-traffic`.
+  >> serve.list_deployments()
+  {'MyFirstDeployment': <ray.serve.api.ServeDeployment object at 0x7fe0c04c42b0>}
```
Comment (simon-mo) [inline doc/source/serve/core-apis.rst:43]: oof, we need a better `__repr__` for ServeDeployment

File: doc/source/serve/core-apis.rst [hunk 4]
```diff
@@ -1,128 +1,150 @@
...
+=====================
+
+Often you want to be able to update your code or configuration options for a deployment over time.
+Deployments can be updated simply by updating the code or configuration options and calling ``deploy()`` again.
 
 .. code-block:: python
 
-  >>> serve.list_endpoints()
-  {'simple_endpoint': {'route': '/simple', 'methods': ['GET'], 'traffic': {}}}
+  @serve.deployment(name="my_deployment", num_replicas=1)
```
Comment (simon-mo) [inline doc/source/serve/core-apis.rst:93]: add comment this is the SAME file over time

File: doc/source/serve/http-servehandle.rst [hunk 5]
```diff
@@ -1,92 +1,142 @@
...
+
+    # ... add more routes, routers, etc. to `app` ...
 
-You can return a `Starlette Response object <https://www.starlette.io/responses/>`_ from your Ray Serve backend code:
+    @serve.deployment(route_prefix="/")
+    @serve.ingress(app)
+    class FastAPIWrapper:
+        pass
+
+This is useful for scaling out an existing FastAPI app with no modifications necessary. You can also combine routes defined this way with routes defined on the Deployment:
```
Comment (simon-mo) [inline doc/source/serve/http-servehandle.rst:95]: add a note saying existing middlewares, docs, and advanced feature, etc should work as-is

File: doc/source/serve/http-servehandle.rst [hunk 6]
```diff
@@ -110,146 +160,85 @@ You can follow the same pattern for other Starlette middlewares.
...
-    handle = serve.get_handle("api_endpoint")
-    ray.get(handle.remote(request))
-
-
-Accessing data from the request
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-
-When the request arrives in the model, you can access the data similarly to how
-you would with an HTTP request. Here are some examples how Ray Serve's built-in 
-``ServeRequest`` mirrors ```starlette.requests.request``:
```
Comment (simon-mo) [inline doc/source/serve/http-servehandle.rst:141]: did we end up deleting `ServeRequest`?

**Review by simon-mo — APPROVED** [review]

File: doc/source/serve/tutorials/batch.rst [hunk 7]
```diff
@@ -88,12 +88,11 @@ Ray Serve was able to evaluate them in batches.
 
 What if you want to evaluate a whole batch in Python? Ray Serve allows you to send
 queries via the Python API. A batch of queries can either come from the web server
-or the Python API. Requests coming from the Python API will have a similar API
-to Starlette Request. See more on the API :ref:`here<serve-handle-explainer>`.
+or the Python API. Learn more :ref:`here<serve-handle-explainer>`.
 
-To query the backend via Python API, we can use ``serve.get_handle`` to receive
+To query the backend via Python API, we can use ``Deployment.get_handle`` to receive
```
Comment (architkulkarni) [inline doc/source/serve/tutorials/batch.rst:93]:
```suggestion
To query the deployment via the Python API, we can use ``Deployment.get_handle`` to receive
```

File: python/ray/serve/examples/doc/aiohttp/aiohttp_deploy_serve.py [hunk 8]
```diff
@@ -9,13 +9,12 @@
 serve.start(http_host=None, detached=True)
 
 
-# Define a function to serve. Alternatively, you could define a stateful class.
+# Set up a deployment with the desired number of replicas. This could also be
+# e a stateful class (e.g., if we had an expensive model to set up).
```
Comment (architkulkarni) [inline python/ray/serve/examples/doc/aiohttp/aiohttp_deploy_serve.py:13]:
```suggestion
# a stateful class (e.g., if we had an expensive model to set up).
```

File: python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py [hunk 9]
```diff
@@ -6,32 +6,28 @@
...
-    global serve_handle
-    serve_handle = serve.get_handle("generate")
+    # Deploy our GPT2 Deployment.
+    GPT2.deploy()
 
 
 @app.get("/generate")
 async def generate(query: str):
-    return await serve_handle.remote(query)
+    # Get a handle to our Ray Serve endpoint so we can query it in Python.
```
Comment (architkulkarni) [inline python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py:31]:
```suggestion
    # Get a handle to our Ray Serve deployment so we can query it in Python.
```

File: python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py [hunk 10]
```diff
@@ -6,32 +6,28 @@
 
 app = FastAPI()
```
Comment (architkulkarni) [inline python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py:7]: Is it worth keeping this "Scaling up a FastAPI application" section around now that we have the tighter FastAPI integration?

**Review by architkulkarni — APPROVED** [review]
Looks great!  Added a few more nits.

File: doc/source/serve/core-apis.rst (hunk 3 above)
Comment (edoakes) [reply doc/source/serve/core-apis.rst:43]: Added! There was already a __str__, I just called into it for repr

File: doc/source/serve/core-apis.rst (hunk 4 above)
Comment (edoakes) [reply doc/source/serve/core-apis.rst:93]: done

File: doc/source/serve/http-servehandle.rst (hunk 6 above)
Comment (edoakes) [reply doc/source/serve/http-servehandle.rst:141]: Deployments get regular args and kwargs now

File: doc/source/serve/http-servehandle.rst [hunk 11]
```diff
@@ -1,92 +1,142 @@
...
+    @serve.ingress(app)
+    class MyFastAPIDeployment:
+        @app.get("/")
+        def root(self):
+            return "Hello, world!"
 
-the ``username`` parameter will be accessible in your backend code as follows:
+    MyFastAPIDeployment.deploy()
+
+Now if you send a request to ``/hello``, this will be routed to the ``root`` method of our Deployment. We can also easily leverage FastAPI to define multiple routes with different HTTP methods:
```
[the remark this answers is not in the document]
Comment (edoakes) [reply doc/source/serve/http-servehandle.rst:55]: yeah we should be consistent. I'll make them all lower case.

File: python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py (hunk 10 above)
Comment (edoakes) [reply python/ray/serve/examples/doc/fastapi/servehandle_fastapi.py:7]: Good question. I think it's fine to keep for now but we might consider updating it to use the new "drop-in" support

**Pull request merged**